<h2 align = "center">Zixi Zhang</h2>
<h2 align = "center">505851064</h2>

## Chapter05

**(7)**&#8193;Find by hand the Cholesky decomposition of the matrix 
$$ A = \begin{pmatrix}
2 & -2 \\
    -2 & {5}\end{pmatrix}$$

$\textit{Solution}$: By defination  
$$A = L \cdot L^T$$  
Let $A = \begin{pmatrix}
a_{11} & a_{12} \\
a_{21} & a_{22}\end {pmatrix}$, and $L = \begin{pmatrix}
b_{11} & 0 \\
b_{21} & b_{22}\end{pmatrix}$, then


$$\begin{align*}
L \cdot L^T
&= \begin{pmatrix}
b_{11} & 0 \\
b_{21} & b_{22}
\end{pmatrix}
\cdot
\begin{pmatrix}
b_{11} & b_{21} \\
0 & b_{22}
\end{pmatrix}\\
&= \begin{pmatrix}
(b_{11})^2 & b_{11} \cdot b_{21} \\
b_{11} \cdot b_{21} & (b_{21})^2+(b_{22})^2
\end{pmatrix}
\end{align*}$$


So we can get  
$$\begin{pmatrix}
(b_{11})^2 & b_{11} \cdot b_{21} \\
b_{11} \cdot b_{21} & (b_{21})^2+\left(b_{22}\right)^2
\end{pmatrix} = \begin{pmatrix}
a_{11} & a_{12} \\
a_{21} & a_{22}\end{pmatrix}$$

Here, $ A =\begin{pmatrix}
2 & -2 \\
-2 & {5}\end{pmatrix}$

So,
$$
(b_{11})^2 = 2\\  
b_{11} \cdot b_{21} = -2\\
(b_{11})^2 + (b_{22})^2 =5
$$

$$b_{11} = \sqrt{2} \\
b_{21} = -\sqrt{2}\\
b_{22} = \sqrt{3}\\
$$

$$L = \begin{pmatrix}
\sqrt{2} & 0 \\
-\sqrt{2} & \sqrt{3}\end{pmatrix}$$

**(13)**&#8193;Find the QR decomposition of the matrix  
$$\begin{pmatrix} 1&3&3\\
1 & 3 & 1\\
1 & 1 & 5\\
1 & 1& 3 \end{pmatrix}$$
by the Gram–Schmidt process.

$\textit{Solution}$:
$\overrightarrow{\mathbf{v_1}}=\left(\begin{array}{c}
1 \\
1 \\
1 \\
1
\end{array}\right), \overrightarrow{\mathbf{v_2}}=\left(\begin{array}{l}
3 \\
3 \\
1 \\
1
\end{array}\right), \overrightarrow{\mathbf{v_3}}=\left(\begin{array}{c}
3 \\
1 \\
5 \\
3
\end{array}\right)$

According to the Gram-Schmidt process, $\overrightarrow{\mathbf{u}_{\mathbf{k}}}=\overrightarrow{\mathbf{v}_{\mathbf{k}}}-\sum_{j=1}^{k-1} \operatorname{proj}_{\overrightarrow{\mathbf{u}}_{\mathbf{j}}}\left(\overrightarrow{\mathbf{v}_{\mathbf{k}}}\right)$, where $\operatorname{proj}_{\overrightarrow{\mathbf{u}}_{\mathbf{j}}}\left(\overrightarrow{\mathbf{v}_{\mathbf{k}}}\right)=$ $\frac{\overrightarrow{\mathbf{u}_j} \cdot \overrightarrow{\mathbf{v}_{\mathbf{k}}}}{\left|\overrightarrow{\mathbf{u}_{\mathbf{j}}}\right|^2} \overrightarrow{\mathbf{u}_{\mathbf{j}}}$ is a vector projection.

$\text { The normalized vector is } \overrightarrow{\mathbf{e}_{\mathbf{k}}}=\frac{\overrightarrow{\mathbf{u}_{\mathbf{k}}}}{\left|\overrightarrow{\mathbf{u}_{\mathbf{k}}}\right|}$

$\overrightarrow{\mathbf{u}_1}=\overrightarrow{\mathbf{v}_1}=\left(\begin{array}{c}
1 \\
1 \\
1 \\
1
\end{array}\right) , \quad 
 \overrightarrow{|\mathbf{u_1}|}=\sqrt{\sum_{i=1}^n\left|u_i\right|^2}=2 , \quad 
\overrightarrow{\mathbf{e}_{\mathbf{1}}}=\frac{\overrightarrow{\mathbf{u}_{\mathbf{1}}}}{\left|\overrightarrow{\mathbf{u}_{\mathbf{1}}}\right|} = \left(\begin{array}{c}
\frac{1}{2} \\
\frac{1}{2} \\
\frac{1}{2} \\
\frac{1}{2}
\end{array}\right)
$

$\operatorname{proj}_{\overrightarrow{\mathbf{u_1}}}(\overrightarrow{\mathbf{v_2}})=\frac{\overrightarrow{\mathbf{v_2}} \cdot \overrightarrow{\mathbf{u_1}}}{|\overrightarrow{\mathbf{u_1}}|^2} \overrightarrow{\mathbf{u_1}} = \frac{8}{2^2} \cdot \left(\begin{array}{c}
1 \\
1 \\
1 \\
1\end{array}\right) = 
\left(\begin{array}{c}
2 \\
2 \\
2 \\
2\end{array}\right)$

$\overrightarrow{\mathbf{u}_{\mathbf{2}}}=\overrightarrow{\mathbf{v}_{\mathbf{2}}}-\operatorname{proj}_{\overrightarrow{\mathbf{u}_1}}\left(\overrightarrow{\mathbf{v}_{\mathbf{2}}}\right)=
\left(\begin{array}{c}
3 \\
3 \\
1 \\
1 \end{array}\right)-
\left(\begin{array}{c}
2 \\
2 \\
2 \\
2 \end{array}\right)=
\left(\begin{array}{c}
1 \\
1 \\
-1 \\
-1
\end{array}\right), \quad \overrightarrow{\mathbf{e}_2}=\frac{\overrightarrow{\mathbf{u}_2}}{\left|\overrightarrow{\mathbf{u}_2}\right|}=\left(\begin{array}{c}
\frac{1}{2} \\
\frac{1}{2} \\
-\frac{1}{2} \\
-\frac{1}{2}
\end{array}\right)$

$\operatorname{proj}_{\overrightarrow{\mathbf{u_1}}}(\overrightarrow{\mathbf{v_3}})=\frac{\overrightarrow{\mathbf{v_3}} \cdot \overrightarrow{\mathbf{u_1}}}{|\overrightarrow{\mathbf{u_1}}|^2} \overrightarrow{\mathbf{u_1}} = \frac{12}{2^2} \cdot \left(\begin{array}{c}
1 \\
1 \\
1 \\
1\end{array}\right) = 
\left(\begin{array}{c}
3 \\
3 \\
3 \\
3\end{array}\right), \quad  \operatorname{proj}_{\overrightarrow{\mathbf{u_1}}}(\overrightarrow{\mathbf{v_3}})=\frac{\overrightarrow{\mathbf{v_3}} \cdot \overrightarrow{\mathbf{u_2}}}{|\overrightarrow{\mathbf{u_2}}|^2} \overrightarrow{\mathbf{u_2}} = \frac{-4}{2^2} \cdot \left(\begin{array}{c}
1 \\
1 \\
-1 \\
-1\end{array}\right) = 
\left(\begin{array}{c}
-1 \\
-1 \\
1 \\
1\end{array}\right)$

$\overrightarrow{\mathbf{u}_3}=\overrightarrow{\mathbf{v}}_3-\operatorname{proj}_{\overrightarrow{\mathbf{u}_1}}\left(\overrightarrow{\mathbf{v}_3}\right)-\operatorname{proj}_{\overrightarrow{\mathbf{u}_2}}\left(\overrightarrow{\mathbf{v}_3}\right)=
\left(\begin{array}{c}
3 \\
1 \\
5 \\
3
\end{array}\right) - \left(\begin{array}{c}
3 \\
3 \\
3 \\
3
\end{array}\right) -\left(\begin{array}{c}
-1 \\
-1 \\
1 \\
1
\end{array}\right)=
\left(\begin{array}{c}
1 \\
-1 \\
1 \\
-1
\end{array}\right), \quad \overrightarrow{\mathbf{e}_3}=\frac{\overrightarrow{\mathbf{u}_3}}{\left|\overrightarrow{\mathbf{u}_3}\right|}=\left(\begin{array}{c}
\frac{1}{2} \\
-\frac{1}{2} \\
\frac{1}{2} \\
-\frac{1}{2}
\end{array}\right)$

So, $Q = \begin{pmatrix} \frac{1}{2} & \frac{1}{2} & \frac{1}{2} \\
\frac{1}{2} & \frac{1}{2} & -\frac{1}{2}\\
\frac{1}{2} & -\frac{1}{2} & \frac{1}{2}\\
\frac{1}{2} & -\frac{1}{2} & -\frac{1}{2} \end{pmatrix}$, $\quad$ $Q^T =\begin{pmatrix} \frac{1}{2} & \frac{1}{2} & \frac{1}{2} & \frac{1}{2} \\
\frac{1}{2} & \frac{1}{2} & -\frac{1}{2} & -\frac{1}{2}\\
\frac{1}{2} & -\frac{1}{2} & \frac{1}{2} & -\frac{1}{2} \end{pmatrix}$

In [6]:
R = [1/2 1/2 1/2 1/2; 1/2 1/2 -1/2 -1/2; 1/2 -1/2 1/2 -1/2]*[1 3 3; 1 3 1; 1 1 5; 1 1 3]

3×3 Matrix{Float64}:
 2.0  4.0   6.0
 0.0  2.0  -2.0
 0.0  0.0   2.0

$R = \begin{pmatrix} \frac{1}{2} & \frac{1}{2} & \frac{1}{2} & \frac{1}{2} \\
\frac{1}{2} & \frac{1}{2} & -\frac{1}{2} & -\frac{1}{2}\\
\frac{1}{2} & -\frac{1}{2} & \frac{1}{2} & -\frac{1}{2} \end{pmatrix} \cdot \begin{pmatrix} 1&3&3\\
1 & 3 & 1\\
1 & 1 & 5\\
1 & 1& 3 \end{pmatrix} = \begin{pmatrix} 2 & 4 & 6 \\
0 & 2 & -2\\
0 & 0 & 2\end{pmatrix}$

**(15)**&#8193;If $X =QR$ is the QR decomposition of X, then show that the projection matrix
$$ X(X^*X)^{-1}X^* = QQ^* $$
Also show that $|det X| = | det R| $when $X$ is square and in general that$$
det(X^*X) =(detR)^2$$

$\textit{Solution}:$ $X =QR$, so
$$X^*X = R^*Q^*QR = R^*R \quad
\textrm{(as Q has orthogonal columns, so } \begin{equation}Q^*Q = I \end{equation})$$


\begin{align*}
X(X^*X)^{-1}X^* &= X(R^*R)^{-1}X^*\\
&=QR(R^*R)^{-1}R^*Q^*\\
&= QRR^{-1}R^{-*} R^*Q^*\\
&= QQ^*
\end{align*}

Since $X =QR$, $$|detX| =|detQR|=|detQ||detR|$$ 
When $X$ is square, $Q$ is an orthogonal matrix, we can get $$|detQ|=\sqrt{(detQ)^2 }= \sqrt{(detQ^*)(detQ)}= \sqrt{(detQ^*Q)}=\sqrt{detI }=\sqrt{1}=1$$
so,$$|detX| = |detR|$$

In general,
$$det(X^*X) = det((QR)^*QR) = det(R^*Q^*QR) = det (R^*R)=(detR)^2
$$


## Chapter06

**(1)**&#8193;On a one-dimensional problem of your choice, implement Newton's method. Check your result using the fzero function of the Julia package Roots.jl.

$\textit{Solution:  }  \text{choose }$ $$f(x) = e^x - x^4$$
$$f'(x)=e^x - 4x^3$$
\begin{align*}
g(x)&= x − \frac{f(x)}{f′(x)}\\
&= x- \frac{ e^x - x^4}{e^x - 4x^3}
\end{align*}

In [7]:
function newton(f::Function, fp::Function, x::Real, tol::Real)
    (xold, xnew) = (x, zero(x))
    for iteration = 1:100
        xnew = xold - f(xold) / fp(xold)
        if abs(f(xnew)) < tol
            return (xnew, iteration)
            end
        xold = xnew
    end
    return (xnew, 100)
end
f(x) = exp(x) - x^4
fp(x) = exp(x) - 4x^3
(x, iterations) = newton(f, fp, 2, 1e-14)
(x, iterations)

(1.4296118247255556, 6)

The root $x = 1.4296118247255556$ is found in is found in $7$ iterations to the given tolerance.

In [8]:
#check the result
using Roots
f(x) = exp(x) - x^4
x1 = find_zero(f, (0, 4), Bisection())
(x1,x1 == x)

(1.4296118247255556, true)

Write a Julia program to solve Lambert's equation $w e^w=x$ by Newton's method for $x>0$. Prove that the iterates are defined by
$$
w_{n+1}=w_n \frac{w_n+\frac{x}{w_n e^{w_n}}}{w_n+1} .
$$
Make the argument that $w_{n+1}>w_n$ when $w_n e^{w_n}<x$ and that $w_{n+1}<w_n$ when $w_n e^{w_n}>x$

$\textit{Solution}:$  
\begin{align*}
f(w)&=w e^w - x\\
f'(w)&=w e^w +e^w\\
w_{n+1}&=w_n - \frac{f\left(w_n\right)}{f^{\prime}\left(w_n\right)}\\
&= w_n-\frac{w_n e^{w_n} - x}{w_n e^{w_n} +e^{w_n}}\\
&= w_n(1-\frac{e^{w_n} - \frac{x}{w_n}}{w_n e^{w_n} +e^{w_n}})\\
&= w_n(\frac{{w_n e^{w_n} +e^{w_n}}}{{w_n e^{w_n} +e^{w_n}}}-\frac{e^{w_n} - \frac{x}{w_n}}{w_n e^{w_n} +e^{w_n}})\\
&= w_n(\frac{w_ne^{w_n}+\frac{x}{w_n}}{w_n e^{w_n} +e^{w_n}})\\
&=w_n \frac{w_n+\frac{x}{w_n e^{w_n}}}{w_n+1}
\end{align*}

$$\frac{w_{n+1}}{w_n}=\frac{w_n+\frac{x}{w_n e^{w_n}}}{w_n+1}$$
When $w_n e^{w_n}<x$,
\begin{align*}
&\frac{x}{w_n e^{w_n}} > 1\\ 
\Rightarrow &w_n+\frac{x}{w_n e^{w_n}} >w_n +1\\
\Rightarrow &\frac{w_n+\frac{x}{w_n e^{w_n}}}{w_n+1}>1\\
\Rightarrow &\frac{w_{n+1}}{w_n}>1\\
\Rightarrow &w_{n+1}>w_n
\end{align*}
Similarly, when $w_n e^{w_n}>x$,
\begin{align*}
&\frac{x}{w_n e^{w_n}} < 1\\ 
\Rightarrow &w_n+\frac{x}{w_n e^{w_n}} <w_n +1\\
\Rightarrow &\frac{w_n+\frac{x}{w_n e^{w_n}}}{w_n+1}<1\\
\Rightarrow &\frac{w_{n+1}}{w_n}<1\\
\Rightarrow &w_{n+1}<w_n
\end{align*}

In [9]:
function newtonlambert( x::Real, tol::Real) 
    x < 0 && error(" x > 0 required ")
    function f(w::Real)
        return w*exp(w) - x
    end
    function fp(w::Real)
        return w*exp(w) +exp(w)
    end
    (xold, xnew) = (3, 0)
    for iteration = 1:100
        xnew = xold - f(xold) / fp(xold)
        if abs(f(xnew)) < tol
            return (xnew, iteration)
            end
        xold = xnew
    end
    return (xnew, 100)
end
(w,iterations) = newtonlambert(2,1e-14)
w

0.8526055020137254

**(5)**&#8193;Consider the function
$$
f(x)= \begin{cases}0 & \text { if } x=0 \\ x+x^2 \sin \left(\frac{2}{x}\right) & \text { if } x \neq 0\end{cases}
$$
Calculate its derivative $f^{\prime}(x)$, and argue that Newton's method tends to be repelled by its root $x=0$. This failure occurs despite the fact that $f(x)$ possesses a bounded derivative in a neighborhood of 0.

$\textit{Solution}:$ 
$$
f^{\prime}(x)  =\begin{cases}0 & \text { if } x=0 \\ 1+2 x \sin \left(\frac{2}{x}\right)-2 \cos \left(\frac{2}{x}\right)&\text { if } x \neq 0\end{cases}
$$

\begin{align*} 
x_{n+1} &= x_n - \frac{f\left(x_n\right)}{f^{\prime}\left(x_n\right)} \\
&= x_n - \frac{x_n+x_n^2 \sin \left(\frac{2}{x_n}\right)}{1+2 x \sin \left(\frac{2}{x_n}\right)-2 \cos \left(\frac{2}{x_n}\right)} \text{  , if } x_n \neq 0
\end{align*}

When $x_n = 0$, $f'(x_n) = 0$, the formula $x_{n+1} = x_n - \frac{f\left(x_n\right)}{f^{\prime}\left(x_n\right)}$ can not hold. So Newton's method tends to be repelled by its root $x=0$.

**(9)**&#8193;For $y$ positive the positive root of the equation $f(x)=\frac{1}{x^2}-y=0$ is $\frac{1}{\sqrt{y}}$. Show that the Newton iterates for finding the root are
$$
x_{n+1}=\frac{x_n\left(3-y x_n^2\right)}{2} \text {. }
$$
Alternatively, $x=\frac{1}{\sqrt{y}}$ solves the equation $g(x)=y x^2-1=0$. Demonstrate that this formulation gives rise to the Newton updates
$$
x_{n+1}=\frac{1}{2}\left(x_n+\frac{1}{y x_n}\right) .
$$
The first scheme involves no reciprocals, but the second scheme has better convergence guarantees. Show that the second scheme satisfies $x_{n+1} \geq \frac{1}{\sqrt{y}}$ regardless of the value of $x_n>0$. Also show that $x_{n+1} \leq x_n$ whenever $x_n \geq \frac{1}{\sqrt{y}}$. Hence, global convergence is assured.

$\textit{Solution}:$  
$\text{for } f(x)=\frac{1}{x^2}-y $
$$f'(x) = \frac{d}{d x}\left(\frac{1}{x^2}\right)-\frac{d}{d x}(y) = -\frac{2}{x^3}$$

\begin{align*} 
x_{n+1} &= x_n - \frac{f\left(x_n\right)}{f^{\prime}\left(x_n\right)} \\
&= x_n - \frac{\frac{1}{x_n^2}-y}{-\frac{2}{{x_n}^3}}\\
&= x_n - (\frac{1}{x_n^2}-y) \cdot(-\frac{x_n^3}{2})\\
&= x_n + \frac{x_n}{2}-\frac{yx_n^3}{2}\\
&= \frac{x_n\left(3-y x_n^2\right)}{2}
\end{align*}

$\text{for } g(x)=y x^2-1$

$$g'(x)=2yx$$
\begin{align*} 
x_{n+1} &= x_n - \frac{g\left(x_n\right)}{g'\left(x_n\right)} \\
&= \frac{yx_n^2-1}{2yx_n}\\
&= \frac{x_n}{2}- \frac{1}{2yx_n}\\
&= \frac{1}{2}\left(x_n+\frac{1}{y x_n}\right)
\end{align*}

**(11)**&#8193;For the weighted least squares criterion
$$
f(\boldsymbol{\beta})=\frac{1}{2} \sum_{i=1}^n w_i\left(y_i-\sum_{j=1}^p x_{i j} \beta_j\right)^2
$$
prove that the minimum is achieved when $\boldsymbol{\beta}=\left(\boldsymbol{X}^* \boldsymbol{W} \boldsymbol{X}\right)^{-1} \boldsymbol{X}^* \boldsymbol{W} \boldsymbol{y}$, where $\boldsymbol{W}$ is a diagonal matrix with $i$ th diagonal entry $w_i>0$.

$\textit{Solution}:$  
\begin{align*}
f(\beta)&=\frac{1}{2}(\mathrm{y}-X \beta)^* W(\mathrm{y}-\mathrm{X} \beta)\\
&=\frac{1}{2}\left(y^* W y-y^* W X \beta-\beta^* X^* W y+\beta^* X^* W X \beta\right)
\end{align*}
The minimum is achieved when $\frac{\partial f(\beta)}{\partial \beta}=0$
\begin{align*}
\frac{\partial f(\beta)}{\partial \beta} &=-\frac{1}{2} y^* W X-\frac{1}{2} y^* W^* X+\frac{2}{2} X^* W X \beta \\
&=-y^* W X + X^* W X \beta\\
&=-X^* W y + X^* W X \beta
\end{align*}
Set this to zero,
\begin{align*}
&X^* W X \beta-X^* W y = 0\\
\Rightarrow &\beta = X^* W X^{-1} X^* W y
\end{align*}

**(12)**&#8193;Demonstrate Woodbury's generalization
$$
\left(\boldsymbol{A}+\boldsymbol{U} \boldsymbol{B} \boldsymbol{V}^*\right)^{-1}=\boldsymbol{A}^{-1}-\boldsymbol{A}^{-1} \boldsymbol{U}\left(\boldsymbol{B}^{-1}+\boldsymbol{V}^* \boldsymbol{A}^{-1} \boldsymbol{U}\right)^{-1} \boldsymbol{V}^* \boldsymbol{A}^{-1}
$$
of the Sherman-Morrison matrix inversion formula for compatible matrices $\boldsymbol{A}$, $\boldsymbol{B}, \boldsymbol{U}$, and $\boldsymbol{V}$.

$\textit{Solution}:$  
Consider block matrix
$X=\left[\begin{array}{cc}
A & U \\
V^* & -B^{-1}
\end{array}\right]$, where $A$ is $n \times n, U$ and $V$ are $n \times k$, and $B$ is $k \times k$.

$$\left[\begin{array}{cc}
A & U \\
V^* & -B^{-1}
\end{array}\right]=\left[\begin{array}{cc}
I & 0 \\
V^* A^{-1} & I
\end{array}\right]\left[\begin{array}{cc}
A & 0 \\
0 & -\left(B^{-1}+V^* A^{-1} U\right)
\end{array}\right]\left[\begin{array}{cc}
I & A^{-1} U \\
0 & I
\end{array}\right]$$

So,
$$\begin{align*}
\left[\begin{array}{cc}
A & U \\
V^* & -B^{-1}
\end{array}\right]^{-1}&=\left[\begin{array}{cc}
I & -A^{-1} U \\
0 & I
\end{array}\right] \cdot\left[\begin{array}{cc}
A^{-1} & 0 \\
0 & -\left(B^{-1}+V^* A^{-1} U\right)^{-1}
\end{array}\right]\left[\begin{array}{cc}
I & 0 \\
-V^* A^{-1} & I
\end{array}\right]\\
& =\left[\begin{array}{cc}
A^{-1}-A^{-1} U\left(B^{-1}+V^* A^{-1} U\right)^{-1} V^* A^{-1} & A^{-1} U\left(B^{-1}+V^* A^{-1}U\right)^{-1} \\
\left(B^{-1}+V^* A^{-1} U\right)^{-1} V^* A^{-1} & -\left(B^{-1}+V^* A^{-1} U\right)^{-1}
\end{array}\right]
\end{align*}$$

Set $ W =\left[\begin{array}{ll}
0 & I \\
I & 0
\end{array}\right]$. So, $W^{-1}=W $

$$W X W=\left[\begin{array}{cc}
-B^{-1} & V^* \\
U & A
\end{array}\right]$$

$$W X^{-1} W=(W X W)^{-1}=\left[\begin{array}{cc}
... & ... \\
... & \left(A+U B V^*\right)^{-1}
\end{array}\right]$$
We have $$\left(A+U B V^*\right)^{-1} = (X^{-1})_{11} = A^{-1}-A^{-1} U\left(B^{-1}+V^* A^{-1} U\right)^{-1} V^* A^{-1}$$
i.e $$\left(A+U B V^*\right)^{-1} = A^{-1}-A^{-1} U\left(B^{-1}+V^* A^{-1} U\right)^{-1} V^* A^{-1}$$